pandas/core/indexes/multi.py

```diff
@@ -1206,6 +1206,12 @@ def from_tuples(cls, tuples, sortorder=None, names=None):
         MultiIndex.from_product : Make a MultiIndex from cartesian product
                                   of iterables
         """
+        if not hasattr(tuples, '__len__'):
+            try:
+                iter(tuples)  # if iterable, then it's iterator or generator
```

Comment:
needs checking in ``from_product`` and ``from_arrays`` as well. 

needs tests that hit all 3 cases